colossus/src/main/scala/colossus/service/ServiceServer.scala

```diff
@@ -205,25 +201,34 @@ with ServerConnectionHandler {
      */
     val response: Callback[O] = if (requestBuffer.size <= requestBufferSize) {
       try {
-        processRequest(request) 
+        processRequest(request)
       } catch {
+        case t: ParseException =>
+          Callback.successful(handleFailure(IrrecoverableError(t)))
         case t: Throwable => {
-          Callback.successful(handleFailure(request, t))
+          Callback.successful(handleFailure(RecoverableError(request, t)))
         }
       }
     } else {
-      Callback.successful(handleFailure(request, new RequestBufferFullException))
+      Callback.successful(handleFailure(RecoverableError(request, new RequestBufferFullException)))
     }
     if (requestMetrics) concurrentRequests.increment()
     response.execute{
       case Success(res) => promise.complete(res)
-      case Failure(err) => promise.complete(handleFailure(promise.request, err))
+      case Failure(err) => promise.complete(handleFailure(RecoverableError(promise.request, err)))
     }
   }
+  
+  override def processBadRequest(reason: Throwable) = {
+    Some(handleFailure(IrrecoverableError(reason)))
+  }
 
-  private def handleFailure(request: I, reason: Throwable): O = {
-    addError(request, reason)
-    processFailure(request, reason)
+  private def handleFailure(error: ProcessingFailure[I]): O = {
+    error match {
+      case RecoverableError(request, reason) => addError(request, reason)
+      case _: IrrecoverableError[I] => // TODO: tick/log w/o request
```

Comment:
I feel like this should be implemented in this PR.  I think you can just have `addError`take an `Option[Input]`